plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -136,23 +141,51 @@ compileLiteral = \case
 -- through it.
 
--- | Get the bytestring content of a string expression, if possible. Follows (Haskell) variable references!
-stringExprContent :: GHC.CoreExpr -> Maybe BS.ByteString
-stringExprContent = \case
-  GHC.Lit (GHC.LitString bs) -> Just bs
-  -- unpackCString# / unpackCStringUtf8# are just wrappers around a literal
-  GHC.Var n `GHC.App` expr
-    | let name = GHC.getName n
-    , name == GHC.unpackCStringName || name == GHC.unpackCStringUtf8Name ->
-      stringExprContent expr
+data StringExprContentAs = AsBytes | AsText
+
+-- | Get the bytestring content of a string expression, if possible.
+-- Follows (Haskell) variable references!
+stringExprContent :: StringExprContentAs -> GHC.CoreExpr -> Maybe BS.ByteString
+stringExprContent contentAs coreExpr = case coreExpr of
+  GHC.Lit (GHC.LitString bytes) ->
+    Just bytes
+  GHC.Var isUnpackCString `GHC.App` GHC.Lit (GHC.LitString bytes)
+    | GHC.getName isUnpackCString == GHC.unpackCStringName ->
+        Just bytes
+  GHC.Var isUnpackCStringUtf8 `GHC.App` GHC.Lit (GHC.LitString bytes)
+    | GHC.getName isUnpackCStringUtf8 == GHC.unpackCStringUtf8Name ->
+    case contentAs of
+      AsText -> Just bytes
+      AsBytes ->
+        -- GHC stores bytestring literals UTF-8 encoded, decoding them at runtime.
+        -- In Plinth we decode such bytestrings in compile-time.
+        BS.pack <$> fromUtf8 (BS.unpack bytes)
   -- See Note [unpackFoldrCString#]
   GHC.Var build `GHC.App` _ `GHC.App` GHC.Lam _ (GHC.Var unpack `GHC.App` _ `GHC.App` expr)
-    | GHC.getName build == GHC.buildName && GHC.getName unpack == GHC.unpackCStringFoldrName -> stringExprContent expr
+    | GHC.getName build == GHC.buildName && GHC.getName unpack == GHC.unpackCStringFoldrName ->
+      stringExprContent contentAs expr
   -- GHC helpfully generates an empty list for the empty string literal instead of a 'LitString'
   GHC.Var nil `GHC.App` GHC.Type (GHC.tyConAppTyCon_maybe -> Just tc)
-    | nil == GHC.dataConWorkId GHC.nilDataCon, GHC.getName tc == GHC.charTyConName -> Just mempty
-  -- Chase variable references! GHC likes to lift string constants to variables, that is not good for us!
-  GHC.Var (GHC.maybeUnfoldingTemplate . GHC.realIdUnfolding -> Just unfolding) -> stringExprContent unfolding
+    | nil == GHC.dataConWorkId GHC.nilDataCon, GHC.getName tc == GHC.charTyConName ->
+      Just mempty
+  -- Chase variable references! GHC likes to lift string constants to variables,
+  -- that is not good for us!
+  GHC.Var (GHC.maybeUnfoldingTemplate . GHC.realIdUnfolding -> Just unfolding) ->
+    stringExprContent contentAs unfolding
+  _ -> Nothing
+
+{- | Decoding that undoes GHC's UTF-8 encoding of bytestring literals:
+
+This isn't a full UTF-8 decoder: it only decodes the subset of UTF-8 that
+is expected to be found in bytestring literals: 0x00 - 0xFF
+-}
+fromUtf8 :: [Word8] -> Maybe [Word8]
```

Comment:
I tried using `Data.Text.Encoding.decodeUtf8` and it didn't work: the error was saying about invalid utf-8 bytes.
It turned out that decoding rules are not super complex so I have implemented them below.

Reply:
I'd probably break my forehead against the wall of trying to reuse GHC stuff, but you really did the right thing here. Great work!

Reply:
This is surprising. An example in the comment would be helpful.

Also, why not fail fast here? That seems better than returning a `Nothing`.

Reply:
> Also, why not fail fast here? That seems better than returning a Nothing.

the `fromUtf8` function is used in the 
```haskell
stringExprContent :: StringExprContentAs -> GHC.CoreExpr -> Maybe BS.ByteString
```
which is not monadic and doesn support our `throwPlain` and `throwSd` functions.

`stringExprContent` is used in pattern positions (via `ViewPatterns`) so failing fast there is associated with the much larger code rewrite.